mlflow/sklearn/__init__.py

```diff
@@ -455,6 +461,265 @@ def load_model(model_uri):
     )
 
 
+def _get_autolog_training_status():
+    return _autolog_training_status
+
+
+class _AutologTrainingStatus:
+    """
+    This class is designed for holding information which is used by autologging metrics
+    It will hold information of:
+    (1) a map of model_id to run_id, this information is registered during `patched_fit`
+    (2) a map of "prediction result object id" to a tuple of dataset name(the dataset is
+       the one which generate the prediction result) and model id.
+    (3) some scope flags which indicates currently in which method scope, we record scope
+       flags for method `model.fit`, `eval_and_log_metrics`, `model.score`.
+       In order to avoid nested/duplicated autologging metric, when run into these scopes,
+       we need temporarily disable the metric autologging.
+    (4) eval_dataset_info_map, it is a double level map:
+       `eval_dataset_info_map[model_uuid][eval_dataset_var_name]` will get a list, each
+       element in the list is a tuple of `(id(eval_dataset_instance), eval_dataset_sample_rows)`
+       This data structure is used for:
+        * generating unique dataset name key when autologging metric
+        * storing the dataset row samples, we need log it into metric_info artifact file.
+    (5) metric_api_call_arg_dict_list_map, it is a map, key is metric api function name,
+       and value is a list of call arguments dict.
+    """
+
+    fit_scope_key = 'fit'
+    eval_and_log_metrics_scope_key = 'eval_and_log_metrics'
+    model_score_scope_key = 'model_score'
+
+    def __init__(self):
+        self.model_id_to_run_id_map = {}
+        self.pred_result_id_to_dataset_name_and_model_id = {}
+        self.scope_flag_dict = {
+            self.fit_scope_key: False,
+            self.eval_and_log_metrics_scope_key: False,
+            self.model_score_scope_key: False
+        }
+        self.eval_dataset_info_map = defaultdict(lambda: defaultdict(list))
+        self.metric_api_call_arg_dict_list_map = defaultdict(list)
+
+    def call_scope(self, scope_key):
+        class CallScope:
+            def __enter__(inner_self):
+                self.scope_flag_dict[scope_key] = True
+
+            def __exit__(inner_self, exc_type, exc_val, exc_tb):
+                self.scope_flag_dict[scope_key] = False
+
+        return CallScope()
```

Comment:
If we can simplify things via https://github.com/mlflow/mlflow/pull/4491/files#r670942918, we could also get rid of `CallScope`.